_source/_posts/2019-12-12-tutorial-kotlin-beginners-guide.md

```diff
@@ -0,0 +1,491 @@
+```
+Unzip the file:
+
+```txt
+unzip kotlin-spring-boot.zip -d kotlin-spring-boot
+cd kotlin-spring-boot
+```
+
+## Secure your Application with OpenID Connect
+If you already have an Okta account, see the Create a Web Application in Okta sidebar below. Otherwise, we created a Maven plugin that configures a free Okta developer account + an OIDC app (in under a minute!).
+
+```txt
+./mvnw com.okta:okta-maven-plugin:setup
+```
+
+You should see the following output:
+
+```
+First name: Jimena
+Last name: Garbarino
+Email address: ***
+Company: ***
+Creating new Okta Organization, this may take a minute:
+OrgUrl: https://dev-123456.okta.com
+Check your email address to verify your account.
+
+Writing Okta SDK config to: /home/indiepopart/.okta/okta.yaml
+
+Configuring a new OIDC, almost done:
+Created OIDC application, client-id: ***
+```
+
+Check your email and follow the instructions to activate your Okta account.
+
+> **If you already have an Okta Developer account**
+> 
+> Log in and create a new Application:
+> - From the Applications page, choose Add Application.
+> - On the Create New Application page, select Web.
+> - Give your app a memorable name, add http://localhost:8080/login/oauth2/code/okta and http://localhost:8080/authorization-code/callback as a Login redirect URIs.
+> 
+> Copy the issuer (found under **API** > **Authorization Servers**), client ID, and client secret into `src/main/resources/application.properties`.
+> 
+```properties
+okta.oauth2.issuer=$issuer
+okta.oauth2.client-id=$clientId
+okta.oauth2.client-secret=$clientSecret
+```
+
+Edit `KotlinSpringBootApplication` located the package `com.okta.developer` to add a controller mapping that will print a welcome message in the browser window.
+
+```kotlin
+package com.okta.developer
+
+import org.springframework.boot.autoconfigure.SpringBootApplication
+import org.springframework.boot.runApplication
+import org.springframework.security.core.annotation.AuthenticationPrincipal
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.RestController
+
+@SpringBootApplication
+@RestController
+class KotlinSpringBootApplication {
+
+    @GetMapping("/")
+    fun hello(@AuthenticationPrincipal authenticationToken: OAuth2AuthenticationToken): String {
+
+        return "Welcome ${authenticationToken.principal.attributes["name"]}"
+    }
+}
+
+fun main(args: Array<String>) {
+    runApplication<KotlinSpringBootApplication>(*args)
+}
+```
+
+As you can see, in the `hello` function, the authenticated user is passed in the `authenticationToken`. `@AuthenticationPrincipal` annotation from Spring Security helps to resolve the principal to the expected principal type. The welcome message is built using **String interpolation**, a Kotlin idiom for variable substitution inside strings.
+
+Run the application with the following command:
+```txt
```

Comment:
```suggestion
```shell
```

Reply:
I've had mixed luck with using `shell`, `sh`, and `bash` most commands end up getting colored wrong, i've given up and started using `txt` _unless_ it's an actual shell script